File: 2016/Rust/aoc_2016/src/day_1.rs
use std::fs;

pub fn day_1_main(){
    part_1();
    part_2();
}

fn part_1(){
    //Ingest data
    let file_name = "inputs/day_1_input.txt";
    let contents = fs::read_to_string(file_name)
        .expect("Failed to read File");

    //Split by each line
    let parts: Vec<&str> = contents.split(", ").collect();

    let mut loc: [i32; 2] = [0, 0];

    //0 = N, 1 = E, 2 = S, 3 = W
    let mut heading: i32 = 0;

    for i in 0..parts.len(){
        //Parse instruction
        let chars = String::from(parts[i].replace("\n", ""));
        let dir_move = chars.chars().next().map(|c| c.to_string()).unwrap();
        let num_move_str:String = chars.chars().skip(1).collect();
        let num_move: i32 = num_move_str.parse()
            .expect("Not a valid number");

        //Change heading
        if dir_move == "R"{
            heading += 1;
        }else{
            heading -= 1;
        }

        //Check for wrap around
        if heading < 0{
            heading = 3;
        }else if heading > 3{
            heading = 0;
        }

        //Move
        if heading == 0{
            loc[0] = loc[0] + num_move;
        }else if heading == 1{
            loc[1] = loc[1] + num_move;
        }else if heading == 2{
            loc[0] = loc[0] -  num_move;
        }else if heading == 3{
            loc[1] = loc[1] - num_move;
        }
    }

    println!("Total distance moved in Part 1: {}", (loc[0].abs() + loc[1].abs()));
}

fn part_2(){
    //Ingest data
    let file_name = "inputs/day_1_input.txt";
    let contents = fs::read_to_string(file_name)
        .expect("Failed to read File");

    //Split by each line
    let parts: Vec<&str> = contents.split(", ").collect();
    
    let mut locmem: Vec<[i32; 2]> = Vec::new();
    let mut loc: [i32; 2] = [0, 0];

    //0 = N, 1 = E, 2 = S, 3 = W
    let mut heading: i32 = 0;

    for i in 0..parts.len(){
        let mut checker: bool = true;

        //Parse instruction
        let chars = String::from(parts[i].replace("\n", ""));
        let dir_move = chars.chars().next().map(|c| c.to_string()).unwrap();
        let num_move_str:String = chars.chars().skip(1).collect();
        let num_move: i32 = num_move_str.parse()
            .expect("Not a valid number");

        //Change heading
        if dir_move == "R"{
            heading += 1;
        }else{
            heading -= 1;
        }

        //Check for wrap around
        if heading < 0{
            heading = 3;
        }else if heading > 3{
            heading = 0;
        }

        //Move
        for _j in 0..num_move{
            if heading == 0{
                loc[0] = loc[0] + 1;
            }else if heading == 1{
                loc[1] = loc[1] + 1;
            }else if heading == 2{
                loc[0] = loc[0] -  1;
            }else if heading == 3{
                loc[1] = loc[1] - 1;
            }

            if locmem.contains(&loc){
                checker = false;
                break;
            }else{
                locmem.push(loc);
            }
        }

        if !checker{
            break;
        }
    }

    println!("Total distance moved in Part 2: {}", (loc[0].abs() + loc[1].abs()));
}

File: 2016/Python/Day_1.py
def parse_input():
    with open("Inputs/day_1_input.txt") as input_file:
        for line in input_file:
           splitter = line.split(", ")

           #Remove newline from the splitter
           splitter[len(splitter)-1] = splitter[len(splitter)-1].replace("\n", "") 

    return splitter

def part_1(list_1):
    loc = [0, 0]

    #0 = N, 1 = E, 2 = S, 3 = W
    heading = 0

    for i in list_1:
        #Parse instruction
        dir_move = i[0]
        num_move = int(i[1:])

        #Change heading
        if dir_move == "R":
            heading += 1
        else:
            heading -= 1

        #Check for wrap around
        if heading < 0:
            heading = 3
        elif heading > 3:
            heading = 0

        #Move
        if heading == 0:
            loc[0] = loc[0] + num_move
        elif heading == 1:
            loc[1] = loc[1] + num_move
        elif heading == 2:
            loc[0] = loc[0] -  num_move
        elif heading == 3:
            loc[1] = loc[1] - num_move

    return abs(loc[0]) + abs(loc[1])


def part_2(list_1):
    loc = [0, 0]
    loc_mem = []

    #0 = N, 1 = E, 2 = S, 3 = W
    heading = 0

    for i in list_1:
        #Parse instruction
        dir_move = i[0]
        num_move = int(i[1:])

        #Change heading
        if dir_move == "R":
            heading += 1
        else:
            heading -= 1

        #Check for wrap around
        if heading < 0:
            heading = 3
        elif heading > 3:
            heading = 0

        #Move
        for j in range(0,num_move):
            if heading == 0:
                loc[0] = loc[0] + 1
            elif heading == 1:
                loc[1] = loc[1] + 1
            elif heading == 2:
                loc[0] = loc[0] -  1
            elif heading == 3:
                loc[1] = loc[1] - 1

            if loc in loc_mem:
                return abs(loc[0]) + abs(loc[1])
            else:
                loc_mem.append(loc.copy())

    print("Couldn't find it...")
    return 0

if __name__ == '__main__':
    # Variables
    list_1 = []

    list_1 = parse_input()

    print("Total Distance: " + str(part_1(list_1)))
    print("Total Distance: " + str(part_2(list_1)))
